src/ImageSharp/Memory/Allocators/Internals/GCAwareConfigurableArrayPool{T}.cs

```diff
@@ -0,0 +1,504 @@
+        {
+            private readonly T[][] buffers;
+            private readonly int poolId;
+            private readonly int numberOfBuffers;
+
+            /// <summary>
+            /// Do not make this readonly; it's a mutable struct
+            /// </summary>
+            private SpinLock spinLock;
+            private int index;
+            private uint firstItemMS;
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="Bucket"/> class.
+            /// </summary>
+            /// <param name="bufferLength">The length of each buffer.</param>
+            /// <param name="numberOfBuffers">The number of buffers each of <paramref name="bufferLength"/>.</param>
+            /// <param name="poolId">The pool id.</param>
+            internal Bucket(int bufferLength, int numberOfBuffers, int poolId)
+            {
+                this.spinLock = new SpinLock(Debugger.IsAttached); // only enable thread tracking if debugger is attached; it adds non-trivial overheads to Enter/Exit
+                this.buffers = new T[numberOfBuffers][];
+                this.BufferLength = bufferLength;
+                this.numberOfBuffers = numberOfBuffers;
+                this.poolId = poolId;
+            }
+
+            /// <summary>Gets an ID for the bucket to use with events.</summary>
+            internal int Id => this.GetHashCode();
+
+            internal int BufferLength { get; }
+
+            internal T[] Rent()
+            {
+                T[][] buffers = this.buffers;
+                T[] buffer = null;
+
+                // While holding the lock, grab whatever is at the next available index and
+                // update the index.  We do as little work as possible while holding the spin
+                // lock to minimize contention with other threads.  The try/finally is
+                // necessary to properly handle thread aborts on platforms which have them.
+                bool lockTaken = false, allocateBuffer = false;
+                try
+                {
+                    this.spinLock.Enter(ref lockTaken);
+
+                    if (this.index < buffers.Length)
+                    {
+                        buffer = buffers[this.index];
+                        buffers[this.index++] = null;
+                        allocateBuffer = buffer == null;
+                    }
+                }
+                finally
+                {
+                    if (lockTaken)
+                    {
+                        this.spinLock.Exit(false);
+                    }
+                }
+
+                // While we were holding the lock, we grabbed whatever was at the next available index, if
+                // there was one. If we tried and if we got back null, that means we hadn't yet allocated
+                // for that slot, in which case we should do so now.
+                if (allocateBuffer)
+                {
+                    if (this.index == 0)
```

Comment:
First of all, this is not thread safe, `this.index` variable is a shared state that must be altered only in locked scope.

`this.index` would never be 0 in this piece of code. If bucket is full (i.e. all buffers are not rented), `this.index` would be equal to zero. After exactly one rent in locked scope `this.index` would be incremented thus at `if(this.index == 0)` line index would always be >= 1.

Reply:
This can be fixed like this:
```csharp
bool lockTaken = false;
int takenIndex = -1;
try
{
    this.spinLock.Enter(ref lockTaken);

    if (this.index < buffers.Length)
    {
        buffer = buffers[this.index];
        buffers[this.index] = null;
        takenIndex = this.index++;
    }
}
finally
{
    if (lockTaken)
    {
        this.spinLock.Exit(false);
    }
}

if (buffer == null)
{
    if (takenIndex == 0)
    {
        // Stash the time the first item was added.
        this.firstItemMS = (uint)Environment.TickCount;
    }
// ...
```

Reply:
Good catch. It's also in the wrong method. Will update.